### ISSUE TITLE
Migrate backend to Python Array API Standard

### ISSUE THREAD

**Issue by Littie28** [HarrisonKramer/optiland#405]
### Checklist

- [x] I have searched the [existing issues](https://github.com/HarrisonKramer/optiland/issues) and [discussions](https://github.com/HarrisonKramer/optiland/discussions) for a similar question or feature request.
- [x] I have read the [documentation](https://optiland.readthedocs.io/en/latest/) and tried to find an answer there.
- [x] I am using the **latest version** of Optiland.
- [x] I have included **all necessary context**.

*Thanks for taking the time to go through this — it really helps us help you!*

### Feature Request

**Is your feature request related to a problem? Please describe.**

> **Related to:** #360 

The current backend abstraction lacks proper type safety and IDE support, making development more difficult than it needs to be. The dynamic dispatch pattern using `__getattr__` in `optiland/backend/__init__.py` prevents type checkers from understanding what functions return and breaks IDE autocomplete. This requires maintaining 157 lines of manual `.pyi` stubs that can easily get out of sync with the actual implementations.

This issue specifically addresses the backend module's type safety problems, which are structural and cannot be solved by simply adding type hints (as proposed in the related issue). The Array API migration would complement the broader typing effort by providing a properly typed foundation for the backend layer.

**Example of the problem:**
```python
import optiland.backend as be

be.set_backend('torch')
x = be.array([1.0, 2.0, 3.0])
# Type checker sees: NDArray | Tensor
# But we know it's Tensor - type narrowing doesn't work

result = be.sin(x)
# IDE autocomplete is unreliable
# Type: NDArray | Tensor (should preserve as Tensor!)
```

Additionally, every new backend function requires updating three files (numpy_backend.py, torch_backend.py, and __init__.pyi), which is error-prone and time-consuming.

**Describe the solution you'd like**

Migrate to the [Python Array API Standard](https://data-apis.org/array-api/latest/) using [`array-api-compat`](https://github.com/data-apis/array-api-compat) and [`types-array-api`](https://github.com/34j/types-array-api). This would provide:

1. **Proper type preservation** using generics:
```python
from array_api._2024_12 import Array
from typing import TypeVar

TArray = TypeVar('TArray', bound=Array)

def compute_sag(x: TArray, y: TArray) -> TArray:
    """Type checker knows: ndarray→ndarray, Tensor→Tensor"""
    return be.sqrt(x**2 + y**2)
```

2. **Native IDE autocomplete** - `array-api-compat` provides typed namespace objects with full IntelliSense support

3. **Zero manual stub maintenance** - Both packages are fully typed and maintained by the Consortium for Python Data API Standards

4. **Multi-backend support** - Easy path to add JAX, CuPy, Dask in the future

**Proposed structure:**
- Core Array API functions (sin, cos, matmul, etc.) delegate to `array-api-compat`
- Optics-specific functions (get_bilinear_weights, histogram2d, etc.) go in `optiland.backend.extensions` module
- PyTorch-specific features (gradients) in `optiland.backend.grad` module

Since Optiland is currently v0.5.7 (beta), this would be an ideal breaking change to make before 1.0.

**Describe alternatives you've considered**

1. **Keep current approach** - Type safety and IDE support remain poor, maintenance burden continues

2. **Improve current stubs** - Doesn't solve fundamental `__getattr__` typing issues with type checkers, and manual maintenance burden remains

3. **Wait for PyTorch Array API to mature** - Delays benefits, and `array-api-compat` already handles compatibility differences

**Additional context**

The Array API Standard is being adopted across the scientific Python ecosystem:
- **SciPy**: Experimental Array API support since v1.12
- **scikit-learn**: Experimental Array API support since v1.0
- **PyTorch**: Active Array API implementation (minimum v1.13)
- **JAX, CuPy**: Full Array API support

**Benefits summary:**

| Current | With Array API |
|---------|----------------|
| ❌ Manual 157-line .pyi stub | ✅ Zero stub maintenance |
| ❌ `NDArray \| Tensor` unions everywhere | ✅ `TArray` generics preserve types |
| ❌ IDE autocomplete from stubs (fragile) | ✅ Native typed autocomplete |
| ❌ Type narrowing doesn't work | ✅ Proper type narrowing |
| ❌ Update 3 files per function | ✅ Single implementation |
| ❌ Adding JAX/CuPy = full reimpl | ✅ Works immediately |

**Risks:**
- Breaking changes to backend API (acceptable for beta/pre-1.0)
- Migration effort
- PyTorch Array API marked "experimental" (but `array-api-compat` abstracts this)

Would appreciate feedback on:
1. Should this be targeted
2. Preference for backwards compatibility layer or breaking api changes and clean implementation?

**References:**
- [Array API Standard](https://data-apis.org/array-api/latest/)
- [array-api-compat](https://github.com/data-apis/array-api-compat)
- [types-array-api](https://github.com/34j/types-array-api)
- [Array API Static Typing](https://data-apis.org/array-api/2024.12/design_topics/static_typing.html)
